### PR TITLE
Elliot/add posting rule operator

### PR DESCRIPTION
**Pull request by elliotberdy-athelas**
We need a way of checking if a given procedure remittance has either the same or different qualities as a given payment item for that procedure.

For example:

Check if proc remit has [same check number, different check date, same era id] as a payment item for the procedure. If yes, return true, else false. 

Therefore, the variable will a list of lists, each nested list containing info about how the payment item lines up with the proc remit. For example, if there were two procedures it may look like this:

[["same_check_number", "same_check_date", "different_paid_amount",...], ["same_check_number", "same_check_date", "same_paid_amount",...]].

Then the user will be able to select which conditions they care about (could be one, some, or all).

This operator takes in all of that information and tells you if there is at least one payment item that matches the description for this procedure. 

I have the following test passing locally in normandy as a proof of concept:

def test_exists_posting_rule_engine_only_operator(dbsession: Session):

    # ----------------------------------------------------------- SETUP
    (
        site,
        facility,
        provider,
        user,
        site_address,
        facility_address,
        provider_address,
        patient,
        encounter,
        procedure,
        claim_row,
        csp,
        era,
        claim_recon,
        claim_remittance,
        procedure_remittance,
    ) = setup_procedure_remit_for_rule_testing(dbsession)

    payment_item = basic_payment_item(
        dbsession=dbsession,
        procedure_remittance_id=procedure_remittance.id,
        claim_submission_id=claim_row.id,
        procedure_id=procedure.id,
        billed_amount_in_usd_cents=2000,
        provider_paid_amount_in_usd_cents=1000,
        era_id=era.id,
        updated_by=user,
        created_by=user,
        date_posted=datetime.now(),
        payer_index=1,
    )

    # ----------------------------------------------------------- CREATE RULE

    rule = Rule(
        site_id=site.id,
        created_by=user.id,
        rule_type=RuleType.REMIT_POSTING,
        rule={
            "actions": [
                {
                    "name": "set_procedure_remittance_for_manual_review",
                    "params": {"manual_review_reason": "TESTING"},
                },
            ],
            "conditions": {
                "all": [
                    {
                        "name": "posted_variables",
                        "value": "['same_check_number', 'same_check_date', 'same_paid_amount']",
                        "operator": "exists_posting_rule_engine_only",
                    },
                ],
            },
        },
        is_active=True,
    )
    dbsession.add(rule)
    dbsession.commit()

    # ----------------------------------------------------------- ASSERT RULE APPLIED
    _, updated_remit, applied_rules = apply_posting_rules_and_update_remit(
        session=dbsession,
        procedure_remittance_id=procedure_remittance.id,
        user_id=user.id,
        commit=True,
    )

    assert updated_remit.id == procedure_remittance.id
    assert applied_rules == [rule]
    assert updated_remit.posting_status == ProcedureRemittancePostingStatus.MANUAL_REVIEW_REQUIRED
    dbsession.refresh(procedure_remittance)
    assert (
        procedure_remittance.posting_status
        == ProcedureRemittancePostingStatus.MANUAL_REVIEW_REQUIRED
    )
    assert procedure_remittance.posted == False
    
    
    # ----------------------------------------------------------- CREATE RULE
    
    # delete first rule
    dbsession.delete(rule)

    rule = Rule(
        site_id=site.id,
        created_by=user.id,
        rule_type=RuleType.REMIT_POSTING,
        rule={
            "actions": [
                {
                    "name": "set_procedure_remittance_for_manual_review",
                    "params": {"manual_review_reason": "TESTING"},
                },
            ],
            "conditions": {
                "all": [
                    {
                        "name": "posted_variables",
                        "value": "['same_check_number', 'different_check_date']",
                        "operator": "exists_posting_rule_engine_only",
                    },
                ],
            },
        },
        is_active=True,
    )
    dbsession.add(rule)
    dbsession.commit()
    
    # ----------------------------------------------------------- ASSERT RULE NOT APPLIED
    _, updated_remit, applied_rules = apply_posting_rules_and_update_remit(
        session=dbsession,
        procedure_remittance_id=procedure_remittance.id,
        user_id=user.id,
        commit=True,
    )

    assert updated_remit.id == procedure_remittance.id
    assert applied_rules == []
    assert updated_remit.posting_status == ProcedureRemittancePostingStatus.MANUAL_REVIEW_REQUIRED
    dbsession.refresh(procedure_remittance)
    assert (
        procedure_remittance.posting_status
        == ProcedureRemittancePostingStatus.MANUAL_REVIEW_REQUIRED
    )
    assert procedure_remittance.posted == False

### CODE REVIEW

**Review by kishen-adit — APPROVED** [review]

File: business_rules/operators.py [hunk 1]
```diff
@@ -276,3 +276,13 @@ def shares_exactly_one_element_with(self, other_value):
     @type_operator(FIELD_SELECT_MULTIPLE)
     def shares_no_elements_with(self, other_value):
         return not self.shares_at_least_one_element_with(other_value)
+
+
+    # This operator is only implemented for the posting rule engine at the moment
+    @type_operator(FIELD_SELECT_MULTIPLE)
+    def exists_posting_rule_engine_only(self, other_value):
```
Comment (honDhan) [inline business_rules/operators.py]: can we think of a better name? how will you prevent others from using it if it's only meant to be used by posting engine folks?

File: tests/test_operators.py [hunk 2]
```diff
@@ -190,3 +190,21 @@ def test_shares_no_elements_with(self):
...
+                        contains_all([2, 1, "A"]))
+
+
+    def test_exists_posting_rule_engine_only(self):
+        self.assertTrue(SelectMultipleType([["same_check_number", "same_check_date"], ["different_check_number", "different_check_date"]]).
+                        exists_posting_rule_engine_only('["same_check_number", "same_check_date"]'))
+        self.assertFalse(SelectMultipleType([["same_check_number", "same_check_date"], ["different_check_number", "different_check_date"]]).
+                         exists_posting_rule_engine_only('["same_check_number", "different_check_date"]'))
+        self.assertTrue(SelectMultipleType([["same_check_number", "same_check_date", "different_era_id"], ["different_check_number", "different_check_date, same_era_id"]]).
+                        exists_posting_rule_engine_only('["same_check_number", "different_era_id"]'))
```
Comment (honDhan) [inline tests/test_operators.py]: the frontend will pass this value? something that looks like `str(some_list)`?

**Review by honDhan — COMMENTED** [review]
will stamp after looking at normandy PR also

File: business_rules/operators.py (hunk 1 above)
Comment (elliotberdy-athelas) [reply business_rules/operators.py]: changed it to `compare_state_with_item` - my hope is that this name is only intuitive if you know exactly what it is doing and therefore will not get used accidentally

File: business_rules/operators.py [hunk 3]
```diff
@@ -276,3 +276,15 @@ def shares_exactly_one_element_with(self, other_value):
     @type_operator(FIELD_SELECT_MULTIPLE)
     def shares_no_elements_with(self, other_value):
         return not self.shares_at_least_one_element_with(other_value)
+
+
+    # This operator is only implemented for the posting rule engine at the moment
+    @type_operator(FIELD_SELECT_MULTIPLE)
+    def compare_state_with_item(self, other_value):
```
Comment (honDhan) [inline business_rules/operators.py:283]: add a label that says `(only in Posting Engine)`

**Review by honDhan — APPROVED** [review]

**Pull request merged**